drivers/ethernet/eth_stm32_hal.c

```diff
@@ -221,17 +222,37 @@ static int eth_tx(struct device *dev, struct net_pkt *pkt)
 	tx_config.Length = total_len;
 	tx_config.TxBuffer = tx_buffer_def;
 
-	hal_ret = HAL_ETH_Transmit(heth, &tx_config, ETH_DMA_TRANSMIT_TIMEOUT);
+	/* tx_buffer is allocated on function stack, we need to wait for the transfer to complete */
+	/* So it is not freed before the interrupt happens */
+	hal_ret = HAL_ETH_Transmit_IT(heth, &tx_config);
+
+	if (hal_ret != HAL_OK) {
+		LOG_ERR("HAL_ETH_TransmitIT: failed! ret=%d err=%d", hal_ret, heth->ErrorCode);
+		res = -EIO;
+		goto error;
+	}
+
+	/* Wait for end of TX buffer transmission */
+	if (k_sem_take(&dev_data->tx_int_sem, K_FOREVER) != 0) {
+		LOG_ERR("HAL_ETH_TransmitIT tx_int_sem take error");
+		res = -EIO;
+		goto error;
+	}
+
 #else
+
 	hal_ret = HAL_ETH_TransmitFrame(heth, total_len);
-#endif /* CONFIG_SOC_SERIES_STM32H7X */
 
 	if (hal_ret != HAL_OK) {
-		LOG_ERR("HAL_ETH_Transmit{Frame}: failed!");
+		LOG_ERR("HAL_ETH_TransmitFrame: failed! ret=%d err=%d", hal_ret, heth->ErrorCode);
```

Comment:
I would just write: `HAL_ETH_Transmit: failed!`. The return values and error codes are not useful for the user, since they are not documented anywhere. The developers can see them from debug interface.

Reply:
Done 👍